### ISSUE TITLE
Fine-tune a ResNet-50 to distinguish pneumonia from normal chest X-ray, and report your model’s performance.

### ISSUE THREAD

**Issue by aa23a** [aa23a/res#1]
# Install required packages if not already installed
if (!require(keras)) install.packages("keras")
if (!require(tensorflow)) install.packages("tensorflow")
if (!require(tfdatasets)) install.packages("tfdatasets")
if (!require(reticulate)) install.packages("reticulate")
if (!require(pROC)) install.packages("pROC")
if (!require(caret)) install.packages("caret")

library(keras)
library(tensorflow)
library(tfdatasets)
library(pROC)
library(caret)

# Set the path to your unzipped Kaggle dataset
base_dir <- "PATH/TO/chest_xray"  # <-- CHANGE THIS

# Data generators with augmentation for training
train_datagen <- image_data_generator(
  rescale = 1/255,
  rotation_range = 10,
  width_shift_range = 0.1,
  height_shift_range = 0.1,
  shear_range = 0.1,
  zoom_range = 0.1,
  horizontal_flip = TRUE,
  fill_mode = "nearest"
)

test_datagen <- image_data_generator(rescale = 1/255)

train_generator <- flow_images_from_directory(
  file.path(base_dir, "train"),
  train_datagen,
  target_size = c(224, 224),
  batch_size = 32,
  class_mode = "binary"
)

val_generator <- flow_images_from_directory(
  file.path(base_dir, "val"),
  test_datagen,
  target_size = c(224, 224),
  batch_size = 32,
  class_mode = "binary",
  shuffle = FALSE
)

test_generator <- flow_images_from_directory(
  file.path(base_dir, "test"),
  test_datagen,
  target_size = c(224, 224),
  batch_size = 32,
  class_mode = "binary",
  shuffle = FALSE
)

# Handle class imbalance with class weights
classes <- train_generator$classes
class_weights <- table(classes)
class_weights <- as.list(sum(class_weights) / (2 * class_weights))
names(class_weights) <- c("NORMAL", "PNEUMONIA")

# Load pre-trained ResNet-50 and fine-tune
base_model <- application_resnet50(
  weights = "imagenet",
  include_top = FALSE,
  input_shape = c(224, 224, 3)
)

model <- keras_model_sequential() %>%
  base_model %>%
  layer_global_average_pooling_2d() %>%
  layer_dropout(0.5) %>%
  layer_dense(units = 2, activation = "softmax")

# Freeze base model for transfer learning
freeze_weights(base_model)

model %>% compile(
  loss = "sparse_categorical_crossentropy",
  optimizer = optimizer_adam(lr = 1e-4),
  metrics = c("accuracy")
)

# Train the model
history <- model %>% fit(
  train_generator,
  steps_per_epoch = train_generator$samples %/% train_generator$batch_size,
  epochs = 10,
  validation_data = val_generator,
  validation_steps = val_generator$samples %/% val_generator$batch_size,
  class_weight = class_weights
)

# Unfreeze and fine-tune (optional, for better results)
unfreeze_weights(base_model)
model %>% compile(
  loss = "sparse_categorical_crossentropy",
  optimizer = optimizer_adam(lr = 1e-5),
  metrics = c("accuracy")
)
history_finetune <- model %>% fit(
  train_generator,
  steps_per_epoch = train_generator$samples %/% train_generator$batch_size,
  epochs = 5,
  validation_data = val_generator,
  validation_steps = val_generator$samples %/% val_generator$batch_size,
  class_weight = class_weights
)

# Evaluate on test set
test_generator$reset()
preds <- model %>% predict(test_generator, steps = ceiling(test_generator$samples / test_generator$batch_size))
pred_labels <- apply(preds, 1, which.max) - 1
true_labels <- test_generator$classes

# Metrics
acc <- sum(pred_labels == true_labels) / length(true_labels)
f1 <- F1_Score(true_labels, pred_labels)
roc <- roc(true_labels, preds[,2])
auc <- auc(roc)

cat(sprintf(\"Test Accuracy: %.4f\\n\", acc))
cat(sprintf(\"Test F1 Score: %.4f\\n\", f1))
cat(sprintf(\"Test AUC-ROC: %.4f\\n\", auc))

# Save model
save_model_hdf5(model, \"pneumonia_resnet50_keras.h5\")

# Helper for F1
F1_Score <- function(true, pred) {
  cm <- confusionMatrix(as.factor(pred), as.factor(true))
  return(cm$byClass[\"F1\"][[1]])
}
